File: 백준/Gold/23061. 백남이의 여행 준비/백남이의 여행 준비.py
if __name__ == '__main__':
    N,M = map(int,input().split())
    weights = [0] * (N+1)
    values = [0] * (N+1)

    for i in range(1, N+1):
        weights[i], values[i] = map(int,input().split())

    bags = [0]
    for i in range(1,M+1):
        bags.append(int(input()))

    INT_MIN = -float('inf')

    K = max(bags)
    dp = [0] * ( K + 1)

    dp[0] = 0
    for i in range(1, N+1):
        for k in range(K, 0, -1):
            if k - weights[i] >= 0 and dp[k-weights[i]] != INT_MIN:
                dp[k] = max(dp[k], dp[k-weights[i]] + values[i])

    ans = INT_MIN
    idx = M + 2
    # print(dp)
    for i in range(1,M+1):
        bag = bags[i]
        effi = dp[bag] / bag
        # print(f"dp[{bag}] : {dp[bag]}, effi: {effi}")
        if ans < effi or (ans == effi and idx > i):
            ans = effi
            idx = i



    print(idx)

File: 백준/Gold/23061. 백남이의 여행 준비/백남이의 여행 준비.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	st = new StringTokenizer(br.readLine());
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	
    	weights = new int[N+1];
    	values = new int[N+1];
    	int maxWeight = 0;
    	for(int i = 1; i<=N; i++) {
    		st = new StringTokenizer(br.readLine());
    		weights[i ] = Integer.parseInt(st.nextToken());
    		values[i] = Integer.parseInt(st.nextToken());
    	}
    	
    	bags = new int[M+1];
    	for(int i = 1; i<=M; i++) {
    		bags[i] = Integer.parseInt(br.readLine());
    		maxWeight = Math.max(maxWeight, bags[i]);
    	}
    	
    	dp = new int[maxWeight + 1];
    	for(int i =1; i<=N;i++) {
    		for(int k = maxWeight; k>=1; k--) {
    			if(k - weights[i] >= 0)
    				dp[k] = Math.max(dp[k], dp[k-weights[i]] + values[i]);
    		}
    	}
    	
    	double ans = -1.0;
    	int ret = M + 2;
    	
    	for(int i=1; i<=M;i++) {
    		int bag = bags[i];
    		double effi = (double)dp[bag] / bag;
    		if(ans < effi || (ans == effi && ret > i)) {
    			ans = effi;
    			ret = i;
    		}
    	}
    	
    	System.out.println(ret);
    	
    	
    	
    }
    static int dp[];
    static int weights[], values[];
    static int N,M;
    static StringTokenizer st;
    static int bags[];
}